Sources/Datadog/Core/Persistence/DataBlock.swift

```diff
@@ -0,0 +1,193 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-2020 Datadog, Inc.
+ */
+
+import Foundation
+
+/// Block size binary type
+internal typealias BlockSize = UInt32
+
+/// Block type supported in data stream
+internal enum BlockType: UInt16 {
+    case event = 0x00
+    case headerV1 = 0x01
+}
+
+/// Reported errors while manipulateing data blocks.
+internal enum DataBlockError: Error {
+    case readOperationFailed(streamError: Error?)
+    case invalidByteSequence
+}
+
+/// A data block in defined by its type and a byte sequence.
+///
+/// A block can be serialized in data stream by following TLV format.
+internal struct DataBlock {
+    /// Type describing the data block.
+    let type: BlockType
+
+    /// The data.
+    var data: Data
+
+    /// Returns a Data block in Type-Lenght-Value format.
+    ///
+    /// A block follow TLV with bytes aligned such as:
+    ///
+    ///     +-  2 bytes -+-   4 bytes   -+- n bytes -|
+    ///     | block type | data size (n) |    data   |
+    ///     +------------+---------------+-----------+
+    ///
+    /// - Returns: a data block in TLV.
+    func serialize() -> Data {
+        var buffer = Data()
+        // T
+        withUnsafeBytes(of: type.rawValue) { buffer.append(contentsOf: $0) }
+        // L
+        withUnsafeBytes(of: BlockSize(data.count)) { buffer.append(contentsOf: $0) }
+        // V
+        buffer += data
+        return buffer
+    }
+}
+
+internal extension Data {
+    /// Returns a Data block in Type-Lenght-Value format.
+    ///
+    /// A block follow TLV with bytes aligned such as:
+    ///
+    ///     +-  2 bytes -+-   4 bytes    -+- n bytes  -|
+    ///     | block type | block size (n) | block data |
+    ///     +------------+----------------+------------+
+    ///
+    /// - Parameters:
+    ///   - type: The data type
+    ///   - data: The data
+    /// - Returns: a byte sequence in TLV format.
+    static func block(_ type: BlockType, data: Data) -> Data {
+        return DataBlock(type: type, data: data).serialize()
+    }
+}
+
+/// A block reader can read TLV formatted blocks from a data input.
+///
+///
+internal final class DataBlockReader {
+    /// The input data stream.
+    private let stream: InputStream
```

Comment:
I'm not against using streams instead of file handlers, but I'm missing a good reason for doing this change now, as part of the data format enhancement.

Since SDK `1.0.0` we went through a bunch of unobvious problems and crashes in `FileHandle` APIs. Both our [file writing](https://github.com/DataDog/dd-sdk-ios/blob/63729c909cad2438695adc86f9f2c61103ffe102/Sources/Datadog/Core/Persistence/Files/File.swift#L50-L78) and [file reading](https://github.com/DataDog/dd-sdk-ios/blob/63729c909cad2438695adc86f9f2c61103ffe102/Sources/Datadog/Core/Persistence/Files/File.swift#L94-L121) code is full of patches for different OS and Swift versions, including [Obj-c errors handling in Swift](https://github.com/DataDog/dd-sdk-ios/blob/63729c909cad2438695adc86f9f2c61103ffe102/Sources/Datadog/Core/Persistence/Files/File.swift#L126) added due to ([`FB7743153`](https://openradar.appspot.com/radar?id=4998297672482816)).

I can't say that `InputStream` APIs are the same buggy as `FileHandle` interface, but thinking of all these hard lessons with went through with `FileHandle`, I'm sceptical on changing this core element right now. How about adding it to the backlog and tackling another time? This way we will avoid shipping both, crucial refactoring and important Foundation APIs switch in a single SDK version.

Reply:
Yes, I saw the patches and I was worried of doing any changes in the `File` implementation. The mechanism here is that `FileHandle` reads the entire batch and `InputStream` navigate through the data in memory.
I agree that it's risky tho, it will be best to tackle that separately 👍
Note that it will require changes in `File` to be able to navigate the data

Reply:
Oh, I haven't realised that we still use `FileHandle` to manage file and only use `InputStream` to read data. That said, it's not a revamp, it's an incremental change, so maybe worth doing it. 🤔💭

Reply:
The `DataBlockReader(url:)` initializer is misleading - seem that we don't use it and it can be removed. Then files will be clearly managed through `FH`.